lang/java/reef-utils/src/main/java/org/apache/reef/util/Optional.java

```diff
@@ -122,11 +118,19 @@ public boolean equals(final Object obj) {
 
   @Override
   public int hashCode() {
-    return this.valueHash;
+    if (this.isPresent()) {
+      return this.value.hashCode();
+    } else {
+      return EMPTY_VALUE_HASH;
+    }
   }
 
   @Override
   public String toString() {
-    return this.valueStr;
+    if (this.isPresent()) {
```

Comment:
Consider the ternary operator for this, as it might trigger different compiler / JIT optimizations.